### ISSUE TITLE
Proposal: Remove all forward references in the datamodel

### ISSUE THREAD

**Issue by matthewmueller** [prisma/datamodel-v1.1-feedback#10]
**Proposal**

This builds off of #9 and #7, I'd like to propose removing all forward references in the datamodel. Instead we'd infer relationships based on back references and unique constraints. 

Based on the example in the Readme, here's a diff of the relationships under this proposal:

*one-to-one:*

```diff
type User @db(name: "user") {
  id: ID! @id
- profile: Profile @relation(link: INLINE)
}

type Profile @db(name: "profile") {
  id: ID! @id
- user: User!
+ user: User! @unique
}
```

*one-to-many:*

```diff
type User @db(name: "user") {
  id: ID! @id
-  profile: Profile @relation(link: INLINE)
}

type Profile @db(name: "profile") {
  id: ID! @id
  user: User!
  bio: String!
}
```

*many-to-many:*

```diff
type Post @db(name: "post") {
  id: ID! @id
-  categories: [Category!]! @relation(link: TABLE, name: "PostToCategory")
}

type Category @db(name: "category") {
  id: ID! @id
- posts: [Post!]! @relation(name: "PostToCategory")
}

-type PostToCategory @db(name: "post_to_category") @linkTable {
+type PostToCategory @db(name: "post_to_category") @unique("post", "category") {
  post: Post!
  category: Category!
}
```

**Motivation**

I think the main concern is not being able to generate the client APIs. This is still possible under this proposal. From the perspective of the client libraries, nothing will change. We'll still be able to make APIs that have forward references, though it's a tiny bit more work for the generator scripts.

Additional, I think this approach has a couple advantages:

- It's a more common approach to modeling relationships, less learning and more existing resources online
- Less edge cases, less forgotten links. Avoids user errors and ambiguities like #7

**Potential Solution**

The diffs above show my suggested solution, but there's definitely other ways to syntactically represent these relationships that are semantically equivalent.

**Potential Drawbacks**

- By offering only one way for datamodel to establish each relationship, we may be missing valid use cases for customers. 

My sense on this drawback is there's been enough research into data modeling and schema normalization that we'll probably able to suggest an alternative way to model their data for people that need it, in the same way that we'll be suggesting better ways to do transactions.

- Forward references are more intuitive than back references

I completely agree with this drawback and in #7 I made this mistake without even knowing it. The problem is that under the current spec, you'd still need to know about back references to correctly setup the relationships. 

I'd rather only know I need to setup the one back reference than know I need to setup both the back reference and the forward reference. 

We can also rely on existing literature and current best practices to make sure developers know how to setup their datamodels.

- It will be a breaking change

I'm not sure how important this is, but this would be a pretty major departure from the existing datamodel and implementations. I have a feeling existing prisma users may complain about this change.

**Comment by matthewmueller** [comment]
**Possible Alternatives:**

*1.* Forward references are syntactic sugar, they are optional and don't provide any information about the relationships. Their purpose could be to add clarity to the datamodel file.

*2.* Forward references are optional, don't establish the relationship, but may inform what gets generated. If you don't provide them, you couldn't do `await user(10).posts()`. Essentially, they become a switch.

**Comment by sorenbs** [comment]
I think we could address the `less edge cases` motivation point by implicitly adding a back-relation with the name of the other model. The following datamodels would be equivalent:

```graphql
type User {
  id: ID! @id
  posts: [Post!]!
}

type Post {
  id: ID! @id
  user: User # this field name  comes from the model name
}
```

```graphql
type User {
  id: ID! @id
  posts: [Post!]!
}

type Post {
  id: ID! @id
}
```

**Notes**

- It is possible to customise the name by simply adding the field
- If there are two or more relations between the two models, you have to be explicit. This limitation is already in place.
- I think the default for implicit back relations should be `nullable`

**Drawbacks**

- This is introducing more implicit (magic) behavior

I think we could adopt this in Datamodel v1.1 during the preview period

**Comment by matthewmueller** [comment]
That would also address: 

**3.** Forward references are more intuitive than back references

Basically go further in on forward references and de-emphasize back references. I quite like that! 

I wonder if that concept could also be applied to many-to-many relationships? I guess you often still want a join table so you can add fields that apply to the relationship.

**Comment by sorenbs** [comment]
I don't think we can apply the implicit back relationship to many-to-many relationships, because there is no way to know if the developer wanted a many or single back relation. My suggestion is to always default to a single, optional back-relation-field.

**Comment by matthewmueller** [comment]
Got it! And what's the motivation(s) for wanting to maintain forward references?

**Comment by mavilein** [comment]
@matthewmueller :
I can see how your proposal makes sense for a SQL centric mindset. The motivation for forward references is to have a database agnostic representation. Your proposal replaces forward references in favour of relation/link tables. But SQL databases are the only connectors that support this concept.
We need relation tables in SQL databases because you can't have a list foreign keys in those databases. In more modern databases you can have those and relation tables are not needed.

If we instead stick with our current approach it will work for SQL, Mongo and any future connector that supports relations. The datamodel for each connector will look very similar:

**SQL**:
```graphql
type Post @db(name: "post") {
  id: ID! @id
  categories: [Category!]!
}

type Category @db(name: "category") {
  id: ID! @id
  posts: [Post!]! 
}
```
**Mongo and any connector to come**:
```graphql
type Post @db(name: "post") {
  id: ID! @id
  categories: [Category!]! @relation(link: INLINE)
}

type Category @db(name: "category") {
  id: ID! @id
  posts: [Post!]! 
}
```

**Comment by matthewmueller** [comment]
@mavilein, thank you for clarifying! I think you're right that it's probably wrong to say that there shouldn't be any forward relations, but what if forward relations were used purely for embedded models? Would that work?

I think the use case you've brought up is extremely important to cover. I saw somewhere that Facebook embeds foreign keys as arrays [1] in MySQL relationships to avoid costly lookups across tables at the expense of data normalization (they have these integrity jobs to reduce data inconsistency). Though I haven't used MongoDB in a number of years, I'd imagine this use case applies there too. 

In your example, I think it would be something like this as data:

```js
// post
{
  id: 1,
  categories: [1, 2, 3]
}

// categories
{
  id: 1,
  posts: [1, 8, 9]
}
```

I think this is a pretty awesome feature and definitely think we should support it. I'm just wondering if that could be covered via the following:

```graphql
type Post @db(name: "post") {
  id: ID! @id
  categories: [Category!]!
}

type Category @db(name: "category") {
  id: ID! @id
  posts: [Post!]! 
}
```

I'm probably missing something important about the implementation, so if that's the case I'd love to hear why we need `@relation(link: INLINE)`

**Footnotes:**

**[1]** Not sure how this is actually represented at facebook, but we could use MySQLs `json` type

**Comment by mavilein** [comment]
@matthewmueller :
The data actually looks like the following. The ids are only maintained on the side of the `Post`. Therefore you have to provide the `link: INLINE` so Prisma knows where the references should be stored. 
When you actually navigate the back relation through `Category.posts` Prisma does a "join" to find the related posts. We don't need to maintain the information on both sides to accomplished and it would be also prone to inconsistencies.
```
// post
{
  id: 1,
  categories: [1, 2, 3]
}

// categories
{
  id: 1
}
```
